boa/src/syntax/parser/mod.rs

```diff
@@ -1258,41 +1318,56 @@ impl Parser {
 
     /// <https://tc39.es/ecma262/#prod-MemberExpression>
     fn read_member_expression(&mut self) -> ParseResult {
-        let mut lhs = if self.peek_skip_lineterminator()?.kind == TokenKind::Keyword(Keyword::New) {
-            self.next_skip_lineterminator()?;
+        let mut lhs = if self
+            .peek_skip_lineterminator()
+            .ok_or(ParseError::AbruptEnd)?
+            .kind
+            == TokenKind::Keyword(Keyword::New)
+        {
+            let _ = self
+                .next_skip_lineterminator()
+                .expect("keyword disappeared");
             let lhs = self.read_member_expression()?;
-            self.expect_punc(Punctuator::OpenParen, "expect '('.")?;
+            self.expect_punc(Punctuator::OpenParen, Some("member expression"))?;
             let args = self.read_arguments()?;
             let call_node = Node::Call(Box::new(lhs), args);
 
             Node::New(Box::new(call_node))
         } else {
             self.read_primary_expression()?
         };
-        while let Ok(tok) = self.peek_skip_lineterminator() {
-            match tok.kind {
+        while let Some(tok) = self.peek_skip_lineterminator().cloned() {
```

Comment:
Could we somehow avoid this? It seems that Rust gives a "multiple mutable reference" error when you return a read-only reference from a mutable reference more than once. Is this something the compiler should understand it's safe?